### ISSUE TITLE
Debugging .Net Core is messed up after instrumenting code with a profiler

### ISSUE THREAD

**Issue by mihail-vladov** [dotnet/coreclr#24680]
Hi, I have a problem with debugging .Net Core. The debugger arrowhead pointer (highlighted line) is not positioned to the correct execution line when debugging .Net Core with profiler enabled. Stepping the code line by line advances the pointer based on the initial offset/messed position. When the arrowhead leaves the method the remaining lines are executed at once.

SampleProfiler uses Profiling API to inject several dummy lines at the beginning of the methods by handling ICorProfilerCallback::JITCompilationStarted event. The final code looks like the following:
```
0000:  nop
0001:  nop

0002:  ldc.i4     0 
0004:  brfalse    0008

0005:  nop
0006:  nop
0007:  nop
0008:  nop

...

       <the original code goes here>

...        
```

I have prepared a [simple project](https://github.com/mihail-vladov/Debug-NetCore-Problem) demonstrating the issue. The problem is not reproducible in .NET Framework. You could use the same project to verify that as it is multitargeting.

Please advise on how this problem can be overcome.

**Comment by hoyosjs** [comment]
@davmason 

**cc** @dotnet/dotnet-diag

**Comment by davmason** [comment]
Are you calling [SetILInstrumentedCodeMap](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilerinfo-setilinstrumentedcodemap-method) in your profiler?

**Comment by mihail-vladov** [comment]
Hi, sorry for the late response. Yes, we are calling the SetILInstrumentedCodeMap. However, everything there seems to be fine as the same profiler works without a problem for .Net Framework for years now.
Is there anything different for .Net Core?
If you have any suggestions on how we could find the roots of this problem is welcome.

**Comment by davmason** [comment]
@mihail-vladov Just to set expectations, debugging a profiler instrumented method is not considered a supported feature. We don't make any guarantees that it will work, and have no testing for it.

With all that being said, I'd still like to understand why it is failing on coreclr and am open to fixing it if the fix is relatively safe.

In your sample project, are you able to share the source code for your profiler? I.e. the C++ sources? If so that would aid in debugging greatly.

**Comment by mihail-vladov** [comment]
@davmason Thank you for the clarification.

I am currently trying to find the appropriate way to provide you access to the profile code from a legal point of view. Because the code is owned by the company I work for. I will get back to you probably at the start of the next week.

**Comment by noahfalk** [comment]
@davmason - One option to consider: If @mihail-vladov used VS to capture a dump of the process when the arrow is in the wrong place, then loaded that dump in VS to confirm the arrow is still in the wrong place, you might be able to get some good info about the issue from that.

**Comment by mihail-vladov** [comment]
@davmason Could you please drop me an email at mihail.vladov at progress.com so I can provide you with the necessary data in a private channel. Thank you in advance.

**Comment by davmason** [comment]
Fixed for 3.1
